### ISSUE TITLE
Add collapsible admonition blocks

### ISSUE THREAD

**Issue by gggto** [tommilligan/mdbook-admonish#18]
The mkdocs-material implementation allows for [collapsible blocks](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#collapsible-blocks), a pretty cool feature that allows for example to hide the answer to a question, which is invaluable for teaching material.

Would it be possible to add it to mdbook-admonish?

**Comment by tommilligan** [comment]
This is definitely possible to add - it looks like mkdocs takes the simplest approace of using the details/summary element, so no JS required: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details

As this isn't something I'm personally interested in, I'm not likely to work on it soon. If you're interested in contributing it as a feature, I'm very happy to provide help along the way!

This will need:
- To update the styling for the added tags, to handle the open/close states: https://github.com/tommilligan/mdbook-admonish/blob/c28f2b2fc9eea2c4899d83e436e9b9df61b11d91/compile_assets/scss/admonition.scss#L117
- To update the HTML templates with the details/summary tags, where appropriate: https://github.com/tommilligan/mdbook-admonish/blob/c28f2b2fc9eea2c4899d83e436e9b9df61b11d91/src/lib.rs#L156
- To (maybe?) add a UI element for triggering open/close, and check that the click handling interacts well with linking to the admonition (currently in progress, https://github.com/tommilligan/mdbook-admonish/pull/19)

**Comment by gggto** [comment]
I am pretty interested in implementing that feature. Thanks a lot for all those pointers!

Do you have any idea on how we should represent a collapsible block in the markdown? I was thinking of something like a "special custom-class", like in:
`````md
```admonish note.collapsible
Something
```
`````
However, such a class would need special handling in the `Admonition::html` function, which may be disturbing/unexpected from users. Would that be ok?

**Comment by tommilligan** [comment]
Hmm, that's a good question. In general I can see this heading down the road of more and more configuration options, so something forward extensible would be nice.

mkdocs uses a different fence style, which I don't want to do as I think it's bad design (and would also be a lot of work for us): https://squidfunk.github.io/mkdocs-material/reference/admonitions/#collapsible-blocks

One option would be to pass in a structure config object, such as a json string with options. This has the benefit of being able to hold whatever config we want, but will be very unpleasant to write.

I wonder if a better design overall is for the whole info string to be key/value pairs, such as:

~~~markdown
```admonish type="note" class=".custom.classname" collapse=true title="My custom title"
```
~~~

We could adopt something simple like TOML key value pairs (or a limited subset of simple values we need).

This would be a v2 change but I'm happy to do that for forward extensibility!

**Comment by gggto** [comment]
Yeah, I do agree that the "fence" option doesn't seem very attractive, especially in terms of clarity and discoverability (let alone ease of implementation and commonmark compliance).

However, while the key/value pair approach seems pretty compelling in terms of clarity, I'm a little concerned about ergonomics. In the common case, I'm not too sure about having to write something like:
````markdown
```admonish type="note"
```
````
Also, the `type` key would have a special status since it would be the only mandatory key.

That said, perhaps all this design needs is a small change: 
````markdown
```admonish note class=".custom.classname" collapse=true title="My custom title"
```
````
Now, we can think of what is after the `admonish` keyword as a `(type, config map)` tuple, of sorts. This has the following advantages:
1. Make it clearer that the `type` "key" is *not* optional since it has a different status and must be before the config key/value pairs.
2. Maybe it's a bit closer to html (like a `<note>` block with options, such as `<note class="something somehting">`).
3. Be close to the existing design.
4. Be easier/more ergonomic to write and read (since we don't have to look for the `type` key, we just look at the `first` word).

Also, but that would be for another issue to discuss, it might be interesting to be able to specify new `type`s with this design, something like writing in your book.toml:
```toml
[preprocessor.admonish]
command = "mdbook-admonish"
custom_types = { fancy_note = [ "note",  { class = "fancy", collapse = true, title = "My fancy note" } ] }
```

**Comment by tommilligan** [comment]
It's worth pointing out that the type key is optional - if omitted it defaults to `note`.

I personally don't think the ergonomics of `type="info"` are too bad, but making it the only positional argument is very feasible - you could just check if the first argument contains `=`, and if not, treat it as the type name. Other than that, everything has to be a key value pair.

I like the extensibility to custom styles idea as well, that would be nice

**Comment by tommilligan** [comment]
So for reference, I propose the following as all valid configuration declarations:

```
admonish
admonish info
admonish type="info"
admonish title="Title"
admonish title="Title" type="info"
```

**Comment by schungx** [comment]
Well, you already use anything behind the type as CSS classes.  That means you have whatever is _before_ the type.

So technically speaking you can map non-ASCII symbols for this purpose.

For example:

~~~
```admonish ?note.my_class "Some hidden note"
~~~

That <kbd>?</kbd> indicates collapsible.

Or, if definite symbols only, it can probably also come _after_ the type but _before_ classes:

~~~
```admonish note?.my_class "Some hidden note"
~~~

Personally I think putting it before the type works well to highlight the option.

**Comment by schungx** [comment]
> So for reference, I propose the following as all valid configuration declarations:
> 
> ```
> admonish
> admonish info
> admonish type="info"
> admonish title="Title"
> admonish title="Title" type="info"
> ```

I would hesitate to require putting `title="..."` on every block, especially when options are rare.  You have to take care of the common path.

**Comment by tommilligan** [comment]
> I would hesitate to require putting `title="..."` on every block, especially when options are rare.  You have to take care of the common path.

Oh to be clear, I mean _any_ of the lines above are valid - all the configuration options remain optional. It's just if a title is specified, it's now with the `title=` key.

I dislike the `?` as a config option because it's opaque, and not extensible. As this is a case of 'please add feature x', I think this a good point to think about how adding future features will gel with the current ones.

**Comment by schungx** [comment]
Yes, the proposed new style is more extensible.  You can also make it complete with `id=my_id` and `class=my_class` syntax to add HTML ID's and CSS styles to the block in addition to the old style.  This makes it compatible with code generators which prefer these type of key=value style.

**Comment by tommilligan** [comment]
@gggto FYI I'm going to work now on adding the new configuration syntax for existing features, which can then be extended with an option for collapsibility at the point it is implemented.

**Comment by tommilligan** [comment]
@gggto the internal config refactoring is merged in #24 and #25 - I think it's now in a stable state for you base your changes on.

I haven't yet updated the readme, but the unit tests provide good coverage of the new format: https://github.com/tommilligan/mdbook-admonish/blob/beb640077f1cfd5834059db323dc835f5f790aef/src/config/v2.rs#L117

You should just be able to use TOML's building boolean syntax.

**Comment by tommilligan** [comment]
Released in v1.7.0
